src/Storages/S3Queue/S3QueueFilesMetadata.cpp

```diff
@@ -28,326 +42,875 @@ namespace
-            break;
+            ProfileEvents::increment(ProfileEvents::CannotRemoveEphemeralNode);
+            LOG_DEBUG(log, "Cannot remove {} since session has been expired", zk_node_path);
         }
     }
-    processed_and_failed_files.merge(getProcessingFiles());
-    return processed_and_failed_files;
+    catch (...)
+    {
+        ProfileEvents::increment(ProfileEvents::CannotRemoveEphemeralNode);
+        DB::tryLogCurrentException(__PRETTY_FUNCTION__, "Cannot remove " + zk_node_path);
+    }
+    return false;
 }
 
-std::shared_ptr<zkutil::EphemeralNodeHolder> S3QueueFilesMetadata::acquireLock(zkutil::ZooKeeperPtr zookeeper)
+void S3QueueFilesMetadata::cleanupThreadFunc()
 {
-    UInt32 retry_count = 200;
-    UInt32 sleep_ms = 100;
-    UInt32 retries = 0;
+    /// A background task is responsible for maintaining
+    /// max_set_size and max_set_age settings for `unordered` processing mode.
 
-    while (true)
+    if (shutdown)
+        return;
+
+    try
+    {
+        cleanupThreadFuncImpl();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    if (shutdown)
+        return;
+
+    task->scheduleAfter(generateRescheduleInterval(min_cleanup_interval_ms, max_cleanup_interval_ms));
+}
+
+void S3QueueFilesMetadata::cleanupThreadFuncImpl()
+{
+    auto timer = DB::CurrentThread::getProfileEvents().timer(ProfileEvents::S3QueueCleanupMaxSetSizeOrTTLMicroseconds);
+
+    chassert(max_set_size || max_set_age_sec);
+
+    const bool check_nodes_limit = max_set_size > 0;
+    const bool check_nodes_ttl = max_set_age_sec > 0;
+
+    const auto zk_client = getZooKeeper();
+    auto nodes = zk_client->getChildren(zookeeper_processed_path);
+    if (nodes.empty())
+    {
+        LOG_TEST(log, "A set of nodes is empty");
+        return;
+    }
+
+    const bool nodes_limit_exceeded = nodes.size() > max_set_size;
+    if (!nodes_limit_exceeded && check_nodes_limit && !check_nodes_ttl)
+    {
+        LOG_TEST(log, "No limit exceeded");
+        return;
+    }
+
+    /// Create a lock so that with distributed processing
+    /// multiple nodes do not execute cleanup in parallel.
+    auto ephemeral_node = zkutil::EphemeralNodeHolder::create(zookeeper_cleanup_lock_path, *zk_client, toString(getCurrentTime()));
+    if (!ephemeral_node)
+    {
+        LOG_TEST(log, "Cleanup is already being executed by another node");
+        return;
+    }
+    /// TODO because of this lock we might not update local file statuses on time on one of the nodes.
+
+    struct Node
+    {
+        std::string name;
+        NodeMetadata metadata;
+    };
+    auto node_cmp = [](const Node & a, const Node & b)
+    {
+        if (a.metadata.last_processed_timestamp == b.metadata.last_processed_timestamp)
+            return a.metadata.file_path < b.metadata.file_path;
+        else
+            return a.metadata.last_processed_timestamp < b.metadata.last_processed_timestamp;
```

Comment:
I think we can simply use `std::tie` here
```suggestion
        return std::tie(a.metadata.last_processed_timestamp, a.metadata.file_path) < std::tie(b.metadata.last_processed_timestamp, b.metadata.file_path);
```